adoc/chapters/device_compiler.adoc

```diff
@@ -456,30 +456,58 @@ to influence the code generated by the device compiler.  These attributes are
 all defined in the [code]#+[[sycl::]]+# namespace, with the exception of the
 deprecated attributes described in <<subsec::attributes.deprecated>>.
 
-Each attribute-token defined in this section may be applied at most once for
-any particular kernel or device function.  Applications that decorate a kernel
-or device function with the same attribute-token more than once are ill formed
-and the compiler must issue a diagnostic.  If a kernel or device function is
-decorated with one of these attributes, the decoration must exist on the first
-declaration of the kernel or device function in the translation unit.  Programs
-which fail to do this are ill formed and the compiler must issue a diagnostic.
-If a function is declared with an attribute in one translation unit and the
-same function is declared without the same attribute in another translation
-unit, the program is ill formed and no diagnostic is required.  Applying these
-attributes to any language construct other than those specified in this section
-has implementation-defined effect.
-
-If any of these attributes is applied to a device function that is also
-compiled for the host, it has no effect when the function is compiled for the
-host.
+If one of the attributes defined in this section is applied to a kernel or
+device function, it must be applied to the first declaration of that kernel
+or device function in the translation unit.  Programs which fail to do this are
+ill formed and the compiler must issue a diagnostic.  Redeclarations of the
+kernel or device function in the same translation unit may optionally have the
+same attribute applied (so long as the attribute arguments are the same between
+the declarations), but this is not required.  The attribute remains in effect
+regardless of whether it appears in the redeclaration.
+
+Unless an attribute's description specifically allows it, a kernel or device
+function may not be declared with the more than one instance of the same
+attribute unless all instances have the same attribute arguments.  The compiler
+must issue a diagnostic for programs which violate this requirement.  When two
+or more instances of the same attribute appear on the declaration of a kernel
```

Comment:
It would be interesting if we introduce some attributes like `[[toggle-property-A]] [[toggle-property-A]]`...
P2156 seems just to talk about standard attributes from the C++ which are idempotent.
But it does not say in the general case what is the semantics of repeating attributes and thus whether they are idempotent or not...
Is this sentence important? If not, can we just remove it?
@AaronBallman ?

Reply:
Are you concerned only with this sentence @keryell?

> When two or more instances of the same attribute appear on the declaration of a kernel or device function, the effect is as though a single instance appeared (assuming that all instances have the same attribute arguments).

I agree that we would need to change that sentence if we ever added an attribute like "toggle", but we don't have any attributes like that now.

I was concerned that the spec wasn't clear about what happens when a attribute is duplicated, which is why I added that sentence.  Do you think it's clear enough if we remove that sentence?

Reply:
> I agree that we would need to change that sentence if we ever added an attribute like "toggle", but we don't have any attributes like that now.

I have no idea what you have in mind for toggle, but you may want to be aware of this: http://eel.is/c++draft/dcl.attr#grammar-4.sentence-4

Reply:
>But it does not say in the general case what is the semantics of repeating attributes and thus whether they are idempotent or not...
>Is this sentence important? If not, can we just remove it?
>@AaronBallman ?

For standard attributes, the standard says whether they can be repeated or not and what happens if they are. For example: http://eel.is/c++draft/dcl.attr.deprecated#3 and http://eel.is/c++draft/dcl.attr.deprecated#4

For vendor-specified attributes, it's up to the vendor to say what happens (so long as it conforms to the rest of the requirements of the standard like appertainment, etc).

As for why the sentence in SYCL is important -- definitions are also declarations, so you want to allow sensible things like:
```
[[attr(1)]] void whatever();
[[attr(1)]] void whatever() {} // This should not be disallowed
```
The quoted sentence helps to clarify that the above code is fine and behaves as though there's only one attribute associated with the declaration.